/******************************************************************************
*
* Copyright Saab AB, 2025 (http://safirsdkcore.com)
*
* Created by: Lars Hagström / lars@foldspace.nu
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/
#include "tracer_viewer_main_window.h"
#include "ui_tracer_viewer_main_window.h"
#include "apps_widget.h"
#include "log_widget.h"
#include "tracer_edit_widget.h"
#include "tracer_data_receiver.h"
#include "highlight_widget.h"
#include "highlight_rule.h"

#include <QFileInfo>
#include <QCloseEvent>
#include <QActionGroup>
#include <QAction>
#include <QDateTime>
#include <QDebug>
#include <QFile>
#include <cmath>
#include <QHeaderView>
#include <QLabel>
#include <QMenu>
#include <QMenuBar>
#include <QMessageBox>
#include <QPushButton>
#include <QScrollBar>
#include <QTabWidget>
#include <QTableView>
#include <QTextBrowser>
#include <QToolBar>
#include <QToolButton>
#include <QFileDialog>
#include <QThreadPool>
#include <boost/asio/io_context.hpp>

#include <Safir/Dob/Typesystem/Internal/InternalOperations.h>
#include <Safir/Dob/Typesystem/Serialization.h>
#include <Safir/Dob/Typesystem/Convenience.h>
#include <Safir/Dob/NotOpenException.h>
#include <Safir/Dob/ConnectionAspectMisc.h>

#include <Safir/Utilities/Internal/Expansion.h>

TracerViewerMainWindow::TracerViewerMainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::TracerViewerMainWindow)
    , m_toolBar(new QToolBar(this))
    , m_connectedLabel(new QLabel())
    , m_receivedLabel(new QLabel(tr("Received: 0")))
    , m_droppedLabel(new QLabel(tr("Dropped: 0")))
    , m_bufferLabel(new QLabel(tr("Live buffer: 0%")))
    , m_socketStatusLabel(new QLabel(tr("Tracer: --")))
    , m_isDispatchSignalled()
    , m_dataReceiver(std::make_shared<TracerDataReceiver>())
    , m_ioWork(boost::asio::make_work_guard(m_io))
{
    ui->setupUi(this);
    m_isDispatchSignalled.clear();

    // keep constructor readable – delegate real work to helpers
    ConfigureDockManager();
    BuildUi();

    SetupToolBarActions();
    SetupStatusBar();
    WireSignals();
    StartStatsTimer();
    StartDobConnection();
    SetupAsio();
}

void TracerViewerMainWindow::ConfigureDockManager()
{
    ads::CDockManager::setConfigFlag(ads::CDockManager::DockAreaHasUndockButton, false);
    ads::CDockManager::setConfigFlag(ads::CDockManager::DockAreaHasCloseButton, false);
    ads::CDockManager::setConfigFlag(ads::CDockManager::MiddleMouseButtonClosesTab, true);
    ads::CDockManager::setConfigFlag(ads::CDockManager::TabCloseButtonIsToolButton, true);
    ads::CDockManager::setConfigFlag(ads::CDockManager::FocusHighlighting, true);
}

void TracerViewerMainWindow::BuildUi()
{
    setAcceptDrops(true);

    m_dockManager = new ads::CDockManager(this);
    m_dockManager->setStyleSheet("");
    OnThemeChanged();

    // central welcome text
    auto* label = new QLabel(
        "Welcome to Safir Tracer Viewer. It is an application for viewing trace\n"
        "logs from Safir applications. See the User's Guide for more information.");
    label->setAlignment(Qt::AlignCenter);

    auto* centralDockWidget = new ads::CDockWidget(m_dockManager, "CentralWidget");
    centralDockWidget->setWidget(label);
    centralDockWidget->setFeature(ads::CDockWidget::NoTab, true);
    centralDockWidget->setFeature(ads::CDockWidget::DockWidgetFocusable, false);
    m_centralDockArea = m_dockManager->setCentralWidget(centralDockWidget);

    // apps dock (left)
    m_apps = new AppsWidget(this);
    auto* appsDock = new ads::CDockWidget(m_dockManager, "Tracers");
    appsDock->setWidget(m_apps, ads::CDockWidget::ForceNoScrollArea);
    appsDock->setFeature(ads::CDockWidget::DockWidgetFocusable, false);
    m_leftDockArea = m_dockManager->addDockWidget(ads::LeftDockWidgetArea, appsDock);
    appsDock->toggleViewAction()->setShortcut(QKeySequence(tr("Ctrl+T")));

    // edit dock
    m_editDock = new ads::CDockWidget(m_dockManager, "Edit Tracer");
    auto* leftArea = m_dockManager->addDockWidget(ads::BottomDockWidgetArea, m_editDock, m_leftDockArea);
    m_editDock->setFeature(ads::CDockWidget::DockWidgetFocusable, false);
    m_editDock->closeDockWidget();

    // live-log dock (centre)
    m_liveLog = new LogWidget(m_dataReceiver, this);
    auto* liveDock = new ads::CDockWidget(m_dockManager, "Live Log");
    liveDock->setWidget(m_liveLog, ads::CDockWidget::ForceNoScrollArea);
    m_dockManager->addDockWidget(ads::CenterDockWidgetArea, liveDock, m_centralDockArea);
    liveDock->toggleViewAction()->setShortcut(QKeySequence(tr("Ctrl+L")));

    // highlight rules editor
    m_highlight = new HighlightWidget;
    auto* highlightDock = new ads::CDockWidget(m_dockManager, "Highlight Rules");
    highlightDock->setWidget(m_highlight, ads::CDockWidget::ForceNoScrollArea);
    m_dockManager->addDockWidget(ads::RightDockWidgetArea, highlightDock, m_centralDockArea);
    highlightDock->closeDockWidget();
    highlightDock->toggleViewAction()->setShortcut(QKeySequence(tr("Ctrl+H")));
    m_highlightAction = highlightDock->toggleViewAction();

    ui->menuView->addAction(appsDock->toggleViewAction());
    ui->menuView->addAction(liveDock->toggleViewAction());
    ui->menuView->addAction(highlightDock->toggleViewAction());

    ApplyHighlightRulesToAllLogs();

    m_centralDockArea->parentSplitter()->setSizes({280,400,200});
    m_centralDockArea->parentSplitter()->setStretchFactor(0,0);
    m_centralDockArea->parentSplitter()->setStretchFactor(1,100);
    m_centralDockArea->parentSplitter()->setStretchFactor(2,0);
    leftArea->parentSplitter()->setStretchFactor(0,50);
    leftArea->parentSplitter()->setStretchFactor(1,50);
}

void TracerViewerMainWindow::SetupToolBarActions()
{
    addToolBar(Qt::TopToolBarArea, m_toolBar);
    m_toolBar->setFloatable(false);
    m_toolBar->setMovable(false);
    m_toolBar->addWidget(ui->menubar);

    // highlight rules editor
    m_highlightAction->setIcon(QIcon(":/img/icons/palette.svg"));
    m_highlightAction->setToolTip(tr("Edit highlight rules"));
    m_toolBar->addAction(m_highlightAction);

    // follow toggle
    m_followAction = new QAction(QIcon(":/img/icons/arrow_down.svg"), tr("Follow"), this);
    m_followAction->setCheckable(true);
    m_followAction->setChecked(true);
    m_followAction->setToolTip(tr("Toggle follow mode (auto-scroll to newest)"));
    m_toolBar->addAction(m_followAction);

    // jump-to-end push-button (same icon – active for non-live logs)
    m_jumpToEndAction = new QAction(QIcon(":/img/icons/arrow_down.svg"), tr("Jump to End"), this);
    m_jumpToEndAction->setToolTip(tr("Scroll to newest log entry"));
    m_jumpToEndAction->setVisible(false);
    m_toolBar->addAction(m_jumpToEndAction);

    // clear push-button
    m_clearAction = new QAction(QIcon(":/img/icons/clear.svg"), tr("Clear Buffer"), this);
    m_clearAction->setToolTip(tr("Remove all entries from the live log"));
    m_clearAction->setEnabled(false);            // enabled only when Live Log is focused
    m_toolBar->addAction(m_clearAction);

}

void TracerViewerMainWindow::SetupStatusBar()
{
    ui->statusbar->addPermanentWidget(m_socketStatusLabel);
    ui->statusbar->addPermanentWidget(m_receivedLabel);
    ui->statusbar->addPermanentWidget(m_droppedLabel);
    ui->statusbar->addPermanentWidget(m_bufferLabel);

    m_receivedLabel->setToolTip(tr("Total number of trace log entries received"));
    m_droppedLabel->setToolTip(tr("Number of trace log entries that were dropped before reaching the viewer"));

    m_instanceLabel = new QLabel(tr("SAFIR_INSTANCE: %1")
                                 .arg(Safir::Utilities::Internal::Expansion::GetSafirInstance()));
    ui->statusbar->addPermanentWidget(m_connectedLabel);
    ui->statusbar->addPermanentWidget(m_instanceLabel);

    OnConnectionClosed();
}

void TracerViewerMainWindow::WireSignals()
{
    connect(m_followAction, &QAction::toggled,
            this, [this](bool enabled)
            {
                if (auto* dock = m_dockManager->focusedDockWidget())
                {
                    if (auto* log = qobject_cast<LogWidget*>(dock->widget()))
                    {
                        log->SetFollowModeEnabled(enabled);
                    }
                }
            });
    connect(m_liveLog,      &LogWidget::FollowModeChanged,
            m_followAction, &QAction::setChecked);
    connect(m_clearAction, &QAction::triggered,
            m_liveLog,     &LogWidget::ClearBuffer);

    connect(m_highlight, &HighlightWidget::rulesChanged, this, &TracerViewerMainWindow::ApplyHighlightRulesToAllLogs);

    // Jump-to-end button always scrolls the focused log to the newest entry
    connect(m_jumpToEndAction, &QAction::triggered, this,
            [this]
            {
                if (auto* dock = m_dockManager->focusedDockWidget())
                {
                    if (auto* log = qobject_cast<LogWidget*>(dock->widget()))
                    {
                        log->JumpToLast();
                    }
                }
            });

    // Update toolbar buttons when focus changes:
    connect(m_dockManager, &ads::CDockManager::focusedDockWidgetChanged,
            this, [this](ads::CDockWidget* /*oldDock*/, ads::CDockWidget* nowDock)
            {
                const bool liveFocused = nowDock && nowDock->widget() == m_liveLog;
                m_clearAction->setEnabled(liveFocused);
                ui->actionSave->setEnabled(liveFocused);

                // show Follow toggle only for the live view
                m_followAction->setVisible(liveFocused);

                // show Jump-to-End button only for non-live log widgets
                m_jumpToEndAction->setVisible(!liveFocused);
            });

    connect(m_apps, &AppsWidget::EditTracerClicked,
            this,  &TracerViewerMainWindow::OnEditTracer);
    connect(ui->actionQuit, &QAction::triggered,
            this, []{ QApplication::quit(); });

    connect(ui->actionClose, &QAction::triggered,
            this, &TracerViewerMainWindow::OnCloseCurrentTab);

    connect(ui->actionOpen,  &QAction::triggered,
            this, &TracerViewerMainWindow::OnOpenFile);
    connect(ui->actionSave,  &QAction::triggered,
            this, &TracerViewerMainWindow::OnSaveFile);

    connect(ui->actionGenerateTestData, &QAction::triggered,
            this, &TracerViewerMainWindow::OnGenerateTestData);

    auto* themeGroup = new QActionGroup(this);
    themeGroup->addAction(ui->actionDarkMode);
    themeGroup->addAction(ui->actionLightMode);
    themeGroup->setExclusive(true);

    connect(themeGroup, &QActionGroup::triggered,
            this, &TracerViewerMainWindow::OnThemeChanged);
    connect(ui->actionTouchMode, &QAction::triggered,
            this, &TracerViewerMainWindow::OnThemeChanged);

    m_apps->SetSearchFocus();

    // Initialise toolbar button states based on current focus
    {
        const bool liveFocused = (m_dockManager->focusedDockWidget() &&
                                  m_dockManager->focusedDockWidget()->widget() == m_liveLog);
        m_clearAction->setEnabled(liveFocused);
        ui->actionSave->setEnabled(liveFocused);

        m_followAction->setVisible(liveFocused);
        m_jumpToEndAction->setVisible(!liveFocused);
    }

    // ------------------------------------------------------------------
    // Set initial socket status label (value never changes afterwards)
    // ------------------------------------------------------------------
    m_socketStatusLabel->setText(m_dataReceiver->socketStatusText());
    m_socketStatusLabel->setToolTip(m_dataReceiver->socketStatusTooltip());
}

void TracerViewerMainWindow::StartStatsTimer()
{
    m_statsTimer.setInterval(1000);
    connect(&m_statsTimer, &QTimer::timeout,
            this, &TracerViewerMainWindow::UpdateStats);
    m_statsTimer.start();
}

void TracerViewerMainWindow::StartDobConnection()
{
    QThreadPool::globalInstance()->start([this]
    {
        int instancePart = 0;
        for (;;)
        {
            try
            {
                m_dobConnection.Open(L"SafirTraceView",
                                     std::to_wstring(instancePart),
                                     0,
                                     this,
                                     this);
                break;
            }
            catch (const Safir::Dob::NotOpenException&)
            {
                ++instancePart;
            }
        }

        QMetaObject::invokeMethod(this,
                                  &TracerViewerMainWindow::OnConnectedToDob,
                                  Qt::QueuedConnection);
    });
}

void TracerViewerMainWindow::SetupAsio()
{
    m_ioThread = std::thread([this]{ m_io.run(); });

    using namespace Safir::Dob::Internal::Control;
    m_controlInfoReceiver = std::make_unique<ControlInfoReceiver>(
        m_io,
        [this](std::int64_t incarnationId, std::int64_t /*nodeId*/)
        {
            QMetaObject::invokeMethod(m_dataReceiver.get(),
                                      [this, incarnationId]()
                                      {
                                          m_dataReceiver->setIncarnationFilter(incarnationId);
                                      },
                                      Qt::QueuedConnection);
        });
    m_controlInfoReceiver->Start();
}

// ============================================================================

TracerViewerMainWindow::~TracerViewerMainWindow()
{
    // Cleanly shut down ControlInfo reception
    if (m_controlInfoReceiver)
    {
        m_controlInfoReceiver->Stop();
        m_controlInfoReceiver.reset();
    }

    m_io.stop();
    if (m_ioThread.joinable())
    {
        m_ioThread.join();
    }

    if (m_dataReceiver)
    {
        m_dataReceiver->stop();
    }

    delete ui;

    if (!m_connected)
    {
        //If the connect is still pending we need to kill the program very forcefully, since there is no way
        //to interrupt the Open call.
        std::quick_exit(0);
    }
}

void TracerViewerMainWindow::OnDoDispatch()
{
    if (!m_isDispatchSignalled.test_and_set())
    {
        QMetaObject::invokeMethod(this,
                                  [this]
                                  {
                                      m_isDispatchSignalled.clear();
                                      m_dobConnection.Dispatch();
                                  },
                                  Qt::QueuedConnection);
    }
}

void TracerViewerMainWindow::OnStopOrder()
{
    auto* dock = m_dockManager->findDockWidget("Edit Tracer");
    if (dock != nullptr)
    {
        dock->deleteDockWidget();
    }

    QApplication::quit();
}

void TracerViewerMainWindow::OnConnectedToDob()
{
    const auto connectionName = QString::fromStdWString(Safir::Dob::ConnectionAspectMisc(this->m_dobConnection).GetConnectionName());
    QString msg = tr("Connected as %1").arg(connectionName);

    m_connectedLabel->setText(tr("Connected"));
    m_connectedLabel->setObjectName("ConnectedLabelConnected");
    m_connectedLabel->setToolTip(tr("Connection name: %1").arg(connectionName));
    m_connected = true;

    style()->unpolish(m_connectedLabel);
    style()->polish(m_connectedLabel);

    m_apps->Initialize(m_dobConnection);
}

void TracerViewerMainWindow::OnConnectionClosed()
{
    m_connectedLabel->setText(tr("Not connected"));
    m_connectedLabel->setObjectName("ConnectedLabelDisconnected");
    m_connectedLabel->setToolTip("");
    m_connected = false;

    style()->unpolish(m_connectedLabel);
    style()->polish(m_connectedLabel);

}
namespace
{
    QString readStyleSheet(const QString& path)
    {
        QFile f(path);
        if (!f.exists())
        {
            throw std::logic_error(QString("Stylesheet %1 could not be found").arg(path).toStdString());
        }

        f.open(QFile::ReadOnly | QFile::Text);
        QTextStream ts(&f);
        return ts.readAll();
    }
}

void TracerViewerMainWindow::OnThemeChanged()
{
    QStringList mainStyleSheet;
    QStringList adsStyleSheet;

    if (ui->actionTouchMode->isChecked())
    {
        mainStyleSheet.append("* {font-size:20px;}");
    }

    if (ui->actionDarkMode->isChecked())
    {
        mainStyleSheet.append(readStyleSheet(":qdarkstyle/dark/darkstyle.qss"));
        mainStyleSheet.append(readStyleSheet(":customizations/tweaks-both.qss"));
        mainStyleSheet.append(readStyleSheet(":customizations/tweaks-dark.qss"));

        adsStyleSheet.append(readStyleSheet(":customizations/ads-dark.qss"));
    }
    else
    {
        mainStyleSheet.append(readStyleSheet(":qdarkstyle/light/lightstyle.qss"));
        mainStyleSheet.append(readStyleSheet(":customizations/tweaks-both.qss"));
        mainStyleSheet.append(readStyleSheet(":customizations/tweaks-light.qss"));

        adsStyleSheet.append(readStyleSheet(":customizations/ads-light.qss"));
    }

    qApp->setStyleSheet(mainStyleSheet.join("\n"));
    m_dockManager->setStyleSheet(adsStyleSheet.join("\n"));
}

void TracerViewerMainWindow::OnEditTracer(const std::int64_t instanceId)
{
    if (m_editDock->widget() != nullptr)
    {
        m_editDock->takeWidget()->deleteLater();
    }

    try
    {
        auto edit = new TracerEditWidget(m_dobConnection, instanceId, this);
        m_editDock->setWidget(edit, ads::CDockWidget::ForceNoScrollArea);
        m_editDock->toggleView(true);
    }
    catch (const std::exception& e) //this is really dob exceptions from the Read in the widget
    {
        ui->statusbar->showMessage(tr("Failed to open tracer editor: %1")
                                   .arg(QString::fromUtf8(e.what())),
                                   5000);      // show for 5 seconds
    }
}

void TracerViewerMainWindow::UpdateStats()
{
    m_receivedLabel->setText(tr("Received: %1").arg(m_dataReceiver->receivedCount()));
    m_droppedLabel->setText(tr("Dropped: %1").arg(m_dataReceiver->droppedCount()));


    const std::size_t size = m_liveLog->bufferSize();
    const std::size_t cap  = m_liveLog->bufferCapacity();
    const int pct          = cap ? static_cast<int>(std::round(double(size) * 100.0 / cap)) : 0;

    m_bufferLabel->setText(tr("Live buffer: %1%").arg(pct));
    m_bufferLabel->setToolTip(tr("%1 / %2 entries").arg(size).arg(cap));

    // Highlight label in red when the circular buffer is full
    if (cap && size >= cap)
    {
        m_bufferLabel->setStyleSheet(QStringLiteral("color:red;font-weight:bold;"));
    }
    else
    {
        m_bufferLabel->setStyleSheet(QString());
    }
}
void TracerViewerMainWindow::OnCloseCurrentTab()
{
    if (m_dockManager)
    {
        if (auto* dock = m_dockManager->focusedDockWidget())
        {
            dock->closeDockWidget();
        }
    }
}

void TracerViewerMainWindow::OnSaveFile()
{
    const QString fileName = QFileDialog::getSaveFileName(
                this,
                tr("Save live log to CSV"),
                QStringLiteral("live_log.csv"),
                tr("CSV Files (*.csv);;All Files (*)"));

    if (fileName.isEmpty())
        return;

    if (!m_liveLog->SaveToCsv(fileName))
    {
        QMessageBox::warning(this,
                             tr("Error saving file"),
                             tr("Failed to save '%1'").arg(fileName));
    }
}

void TracerViewerMainWindow::OnOpenFile()
{
    const QString fileName = QFileDialog::getOpenFileName(
                this,
                tr("Open CSV file"),
                QString(),
                tr("CSV Files (*.csv);;All Files (*)"));

    if (fileName.isEmpty())
        return;

    try
    {
        auto* csvWidget = new LogWidget(fileName, this);
        csvWidget->SetHighlightRules(m_highlight->rules());
        AddTab(QFileInfo(fileName).fileName(),
               csvWidget);
    }
    catch (const std::exception& e)
    {
        QMessageBox::warning(this,
                             tr("Error opening file"),
                             tr("Failed to open '%1':\n%2")
                             .arg(fileName,
                                  QString::fromUtf8(e.what())));
    }
}



void TracerViewerMainWindow::AddTab(const QString& title,
                                    QWidget* widget)
{
    auto* const existingDock = m_dockManager->findDockWidget(title);
    if (existingDock != nullptr)
    {
        existingDock->raise();
        widget->deleteLater();
        return;
    }

    auto* dock = new ads::CDockWidget(m_dockManager, title);
    widget->setParent(dock);
    dock->setWidget(widget, ads::CDockWidget::ForceNoScrollArea);
    dock->setFeature(ads::CDockWidget::DockWidgetDeleteOnClose, true);
    m_dockManager->addDockWidget(ads::CenterDockWidgetArea, dock, m_centralDockArea);
}

void TracerViewerMainWindow::ApplyHighlightRulesToAllLogs()
{
    const auto rules = m_highlight->rules();
    if (m_liveLog)
        m_liveLog->SetHighlightRules(rules);

    for (auto* dock : m_dockManager->dockWidgetsMap().values())
        if (auto* log = qobject_cast<LogWidget*>(dock->widget()))
            log->SetHighlightRules(rules);
}

// ---------------------------------------------------------------------------
//  File-menu helper – generate a batch of random log entries for testing
// ---------------------------------------------------------------------------
void TracerViewerMainWindow::OnGenerateTestData()
{
    if (m_liveLog)
        m_liveLog->GenerateTestData();
}

